File: code/go/966.go
// https://leetcode.com/problems/vowel-spellchecker/
func spellchecker(wordlist []string, queries []string) []string {
	dict0 := make(map[string]struct{})
	dict1 := make(map[string]string)
	dict2 := make(map[string]string)

	normalize := func(s string) string {
		return strings.Map(func(r rune) rune {
			switch r {
			case 'a', 'e', 'i', 'o', 'u':
				return '*'
			default:
				return r
			}
		}, s)
	}

	for _, word := range wordlist {
		dict0[word] = struct{}{}
		val1 := strings.ToLower(word)
		if _, ok := dict1[val1]; !ok { dict1[val1] = word }
		val2 := normalize(val1)
		if _, ok := dict2[val2]; !ok { dict2[val2] = word }
	}

	res := make([]string, len(queries))
	for i, query := range queries {
		if _, ok := dict0[query]; ok {
			res[i] = query
			continue
		}
		val1 := strings.ToLower(query)
		if v, ok := dict1[val1]; ok {
			res[i] = v
			continue
		}
		val2 := normalize(val1)
		if v, ok := dict2[val2]; ok {
			res[i] = v
		} else {
			res[i] = ""
		}
	}
	return res
}

File: code/python/966.py
# https://leetcode.com/problems/vowel-spellchecker/
class Solution:
    def spellchecker(self, wordlist: List[str], queries: List[str]) -> List[str]:
        vowel_map = str.maketrans('aeiou', '*****')
        dict0 = set(wordlist)
        dict1, dict2 = {}, {}
        for word in wordlist:
            word_lower = word.lower()
            dict1.setdefault(word_lower, word)
            dict2.setdefault(word_lower.translate(vowel_map), word)

        res = []
        for query in queries:
            if query in dict0:
                res.append(query)
                continue
            query_lower = query.lower()
            if query_lower in dict1:
                res.append(dict1[query_lower])
                continue
            query_mask = query_lower.translate(vowel_map)
            res.append(dict2.get(query_mask, ''))
        return res
